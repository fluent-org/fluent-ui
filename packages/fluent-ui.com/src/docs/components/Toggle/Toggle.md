---
title: Toggle
components: Toggle
type: Inputs
langKey: "en"
---

# Toggle

<p class="description">Toggles represent a physical switch that allows users to turn things on or off. Use Toggles to present users with two mutually exclusive options (like on/off), where choosing an option results in an immediate action.</p>

## Controlled

```jsx
() => {
  const [checked, set] = React.useState(false)
  function handleChange(c) {
    set(c)
  }
  return (
    <Toggle checked={checked} onChange={handleChange} />
  )
}
```

## WithLabel

```jsx
() => {
  const [checked, set] = React.useState(false)
  function handleChange(c) {
    set(c)
  }
  return (
    <FormLabel label="Toggle">
      <Toggle checked={checked} onChange={handleChange} />
    </FormLabel>
  )
}
```

## Disabled

```jsx
<>
  <Toggle disabled />
  <Toggle disabled checked />
</>
```

