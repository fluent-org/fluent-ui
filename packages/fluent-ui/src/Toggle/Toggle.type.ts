import * as React from 'react'
import * as PropTypes from 'prop-types'
import { StandardProps, StyledProps } from '..'

export type ToggleClassProps =
  | 'root'
  | 'checked'
  | 'disabled'
  | 'checkedAndDisabled'
  | 'circle'
  | 'circleChecked'
  | 'circleDisabled'
  | 'circleCheckedAndDisabled'

export interface ToggleProps
  extends StandardProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement, 'onChange'>,
    StyledProps {
  checked?: boolean
  value?: string
  onChange?: (checked: boolean) => void
  disabled?: boolean
  children?: React.ReactNode
}

export const TogglePropTypes = {
  checked: PropTypes.bool,
  value: PropTypes.string,
  onChange: PropTypes.func,
  disabled: PropTypes.bool,
  children: PropTypes.node
}
