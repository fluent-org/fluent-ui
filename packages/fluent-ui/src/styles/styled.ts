import styled from 'styled-components'
import { th, variant, system, breakpoints } from '@xstyled/system'

export { styled, system, th, variant, breakpoints }

export * from 'styled-components'
export * from '@xstyled/system'
// export * from '@xstyled/prop-types'

export * from './stylize'
