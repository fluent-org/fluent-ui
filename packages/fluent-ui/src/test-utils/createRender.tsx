import * as React from 'react'
import * as PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

interface WrapperProps {
  children: React.ReactNode
}

const Wrapper: React.FC<WrapperProps> = ({ children }): React.ReactElement => (
  <ThemeProvider theme={{}}>{children}</ThemeProvider>
)

Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui: React.ReactElement, options: any) =>
  render(ui, { wrapper: Wrapper, ...options })

export { customRender as render }
