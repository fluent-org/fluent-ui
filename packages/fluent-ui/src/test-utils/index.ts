export * from '@testing-library/react'
// @ts-ignore
export * from './createRender'
