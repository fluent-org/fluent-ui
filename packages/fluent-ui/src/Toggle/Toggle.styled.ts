import { styled, css, th, variant } from '../styles/styled'

export const StyledLabel = styled.label<{ disabled?: boolean }>`
  display: inline-flex;
  justify-content: space-between;
  align-items: center;
  vertical-align: middle;
  cursor: ${({ disabled }): string => (disabled ? 'not-allowed' : 'pointer')};
`

export const StyledLabelText = styled.span`
  font-size: 0.8em;
  margin-left: ${({ children }): string => (children ? '4px' : '0')};
`

const circleSize = 12

export const StyledToggleCircle = styled.span<{
  checked?: boolean
  disabled?: boolean
}>`
  width: ${circleSize}px;
  height: ${circleSize}px;
  margin-top: -${circleSize / 2}px;
  border-radius: 50%;
  position: absolute;
  top: 50%;
  transition: ${th.transition('toggle')};
  ${variant({
    prop: 'checked',
    default: false,
    variants: {
      true: css`
        transform: translate3d(-100%, 0, 0);
        left: 100%;
        margin-left: -3px;
        background-color: ${th.color('white.default')};
      `,
      false: css`
        left: 3px;
        background-color: ${th.color('black.default')};
      `
    }
  })}
  ${variant({
    prop: 'disabled',
    default: false,
    variants: {
      true: css`
        background-color: ${({ checked }: { checked?: boolean; disabled?: boolean }): string =>
          checked ? th.color('white.default') : th.color('standard.dark1')};
      `,
      false: css``
    }
  })}
`

const checked = variant({
  prop: 'checked',
  default: false,
  variants: {
    true: css`
      color: ${th.color('white.default')};
      background-color: ${th.color('primary.default')};
      border-color: ${th.color('primary.default')};
      &:active {
        background-color: ${th.color('primary.dark1')};
        border-color: ${th.color('primary.dark1')};
      }
    `,
    false: css`
      background-color: transparent;
      border-color: ${th.color('black.default')};
      &:active {
        background-color: ${th.color('standard.default')};
        border-color: ${th.color('standard.default')};
      }
    `
  }
})

const disabled = variant({
  prop: 'disabled',
  default: false,
  variants: {
    true: css`
      background-color: ${({ checked }: { checked?: boolean; disabled?: boolean }): string =>
        checked ? th.color('standard.default') : ''};
      border-color: ${th.color('standard.default')};
      pointer-events: none;
    `,
    false: css``
  }
})

export const StyledToggleWrapper = styled.div<{
  checked?: boolean
  disabled?: boolean
}>`
  width: 40px;
  height: 20px;
  border-radius: 20px;
  position: relative;
  border: 2px solid;
  transition: ${th.transition('toggle')};

  ${checked}
  ${disabled}
`

export const StyledToggle = styled.input`
  width: 100%;
  height: 100%;
  opacity: 0;
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  cursor: ${({ disabled }): string => (disabled ? 'not-allowed' : 'pointer')};
`
