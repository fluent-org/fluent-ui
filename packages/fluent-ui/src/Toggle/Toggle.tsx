import * as React from 'react'
import classNames from 'classnames'
import { createUseStyles } from '@fluent-ui/styles'
import { ToggleClassProps, ToggleProps, TogglePropTypes } from './Toggle.type'
import { Theme } from '../styles'
import { styles } from './Toggle.styled'

export const name = 'Toggle'

const useStyles = createUseStyles<Theme, ToggleClassProps>(styles, { name })

const Toggle: React.FC<ToggleProps> = React.forwardRef<HTMLInputElement, ToggleProps>(
  (props, ref): React.ReactElement => {
    const {
      as: Component = 'div',
      className: classNameProp,
      style,
      checked,
      value,
      onChange,
      disabled,
      ...rest
    } = props
    const classes = useStyles(props)
    const className = classNames(
      classes.root,
      {
        [classes.checked]: checked,
        [classes.disabled]: disabled,
        [classes.checkedAndDisabled]: checked && disabled
      },
      classNameProp
    )
    const circleClassName = classNames(classes.circle, {
      [classes.circleChecked]: checked,
      [classes.circleDisabled]: disabled,
      [classes.circleCheckedAndDisabled]: checked && disabled
    })

    function handleChange(e: React.ChangeEvent<HTMLInputElement>): void {
      onChange && onChange(e.target.checked)
    }
    return (
      <Component className={className} style={style}>
        <div className={circleClassName} />
        <input
          ref={ref}
          type="checkbox"
          checked={checked}
          value={value}
          onChange={handleChange}
          disabled={disabled}
          {...rest}
        />
      </Component>
    )
  }
)

Toggle.displayName = `F${name}`

Toggle.propTypes = TogglePropTypes

export default Toggle
