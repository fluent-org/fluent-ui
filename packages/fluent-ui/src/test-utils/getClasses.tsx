import * as React from 'react'
import { createUseStyles } from '@fluent-ui/styles'
import { Styles } from 'jss'
import { render } from './createRender'

export function getClasses<T extends string>(styles: Styles, name: string): Record<T, string> {
  const useStyles = createUseStyles<T>(styles, { name })
  let classes: Record<T, string>
  function Util(props: any): React.ReactElement {
    classes = useStyles(props)
    return <div />
  }
  render(<Util />)
  // @ts-ignore
  return classes
}
