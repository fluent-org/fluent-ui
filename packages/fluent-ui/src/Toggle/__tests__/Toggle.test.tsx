import * as React from 'react'
import { fireEvent } from '@testing-library/react'
import { getClasses, render } from '../../test-utils'
import '@testing-library/jest-dom/extend-expect'

import Toggle, { name } from '../Toggle'
import { styles } from '../Toggle.styled'
import { ToggleClassProps } from '../Toggle.type'

const classes = getClasses<ToggleClassProps>(styles, name)

describe('Toggle', (): void => {
  test('should be support ref', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    const { container } = render(<Toggle ref={ref} />)
    const { current: element } = ref
    const ToggleElement = container.querySelector(`input[type='checkbox']`) as HTMLInputElement
    expect(element).toEqual(ToggleElement)
  })

  test('should be controlled', async (): Promise<void> => {
    const ref = React.createRef<HTMLInputElement>()
    let checked = false
    function handleChange(c: boolean): void {
      checked = c
    }
    render(<Toggle ref={ref} checked={checked} onChange={handleChange} />)
    fireEvent.click(ref.current!)
    expect(checked).toEqual(true)
  })

  test('should be render a disabled Toggle', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    const { container } = render(<Toggle ref={ref} disabled={true} />)
    expect(container.firstChild).toHaveClass(classes.disabled)
    expect(ref.current!.disabled).toEqual(true)
  })
})
