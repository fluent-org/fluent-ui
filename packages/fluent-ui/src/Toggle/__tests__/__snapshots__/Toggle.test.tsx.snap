// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle basic 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
  cursor: pointer;
}

.c4 {
  font-size: 0.8em;
  margin-left: 0;
}

.c2 {
  width: 12px;
  height: 12px;
  margin-top: -6px;
  border-radius: 50%;
  position: absolute;
  top: 50%;
  -webkit-transition: all 250ms cubic-bezier(0.76,0.24,0.23,0.94) 0ms;
  transition: all 250ms cubic-bezier(0.76,0.24,0.23,0.94) 0ms;
  left: 3px;
  background-color: #000000;
}

.c1 {
  width: 40px;
  height: 20px;
  border-radius: 20px;
  position: relative;
  border: 2px solid;
  -webkit-transition: all 250ms cubic-bezier(0.76,0.24,0.23,0.94) 0ms;
  transition: all 250ms cubic-bezier(0.76,0.24,0.23,0.94) 0ms;
  background-color: transparent;
  border-color: #000000;
}

.c1:active {
  background-color: #dedede;
  border-color: #dedede;
}

.c3 {
  width: 100%;
  height: 100%;
  opacity: 0;
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  cursor: pointer;
}

<label
  className="c0"
>
  <div
    className="c1"
  >
    <span
      className="c2"
    />
    <input
      className="c3"
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <span
    className="c4"
  />
</label>
`;

exports[`Toggle prop disabled 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
  cursor: not-allowed;
}

.c4 {
  font-size: 0.8em;
  margin-left: 0;
}

.c2 {
  width: 12px;
  height: 12px;
  margin-top: -6px;
  border-radius: 50%;
  position: absolute;
  top: 50%;
  -webkit-transition: all 250ms cubic-bezier(0.76,0.24,0.23,0.94) 0ms;
  transition: all 250ms cubic-bezier(0.76,0.24,0.23,0.94) 0ms;
  left: 3px;
  background-color: #000000;
  background-color: #b5b5b5;
}

.c1 {
  width: 40px;
  height: 20px;
  border-radius: 20px;
  position: relative;
  border: 2px solid;
  -webkit-transition: all 250ms cubic-bezier(0.76,0.24,0.23,0.94) 0ms;
  transition: all 250ms cubic-bezier(0.76,0.24,0.23,0.94) 0ms;
  background-color: transparent;
  border-color: #000000;
  border-color: #dedede;
  pointer-events: none;
}

.c1:active {
  background-color: #dedede;
  border-color: #dedede;
}

.c3 {
  width: 100%;
  height: 100%;
  opacity: 0;
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  cursor: not-allowed;
}

<label
  className="c0"
  disabled={true}
>
  <div
    className="c1"
    disabled={true}
  >
    <span
      className="c2"
      disabled={true}
    />
    <input
      className="c3"
      disabled={true}
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <span
    className="c4"
  />
</label>
`;
